src/cdomains/arrayDomain.ml

```diff
@@ -941,7 +1003,762 @@ struct
   let to_yojson (x, y) = `Assoc [ (Base.name (), Base.to_yojson x); ("length", Idx.to_yojson y) ]
 end
 
-module AttributeConfiguredArrayDomain(Val: LatticeWithSmartOps) (Idx:IntDomain.Z):S with type value = Val.t and type idx = Idx.t =
+module HelperFunctionsIndexMustMaySets =
+struct
+  module MustSet = SetDomain.Reverse (SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All indexes" end))
+  module MaySet = SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All indexes" end)
+
+  let compute_set len =
+    List.init (Z.to_int len) (Fun.id)
+    |> List.map Z.of_int
+    |> MustSet.of_list
+
+  let must_nulls_remove i must_nulls_set min_size =
+    if MustSet.is_bot must_nulls_set then
+      MustSet.remove i (compute_set min_size)
+    else
+      MustSet.remove i must_nulls_set
+
+  let must_nulls_filter cond must_nulls_set min_size =
+    if MustSet.is_bot must_nulls_set then
+      MustSet.filter cond (compute_set min_size)
+    else
+      MustSet.filter cond must_nulls_set
+
+  let must_nulls_min_elt must_nulls_set =
+    if MustSet.is_bot must_nulls_set then
+      Z.zero
+    else
+      MustSet.min_elt must_nulls_set
+
+  let may_nulls_remove i may_nulls_set max_size =
+    if MaySet.is_top may_nulls_set then
+      MaySet.remove i (compute_set max_size)
+    else
+      MaySet.remove i may_nulls_set
+
+  let may_nulls_filter cond may_nulls_set max_size =
+    if MaySet.is_top may_nulls_set then
+      MaySet.filter cond (compute_set max_size)
+    else
+      MaySet.filter cond may_nulls_set
+
+  let may_nulls_min_elt may_nulls_set =
+    if MaySet.is_top may_nulls_set then
+      Z.zero
+    else
+      MaySet.min_elt may_nulls_set
+end
+
+module NullByte (Val: LatticeWithNull) (Idx: IntDomain.Z): Str with type value = Val.t and type idx = Idx.t =
+struct
+  module MustNulls = SetDomain.Reverse (SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All Null" end))
+  module MayNulls = SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All Null" end)
+  (* (Must Null Set, May Null Set, Array Size) *)
+  include Lattice.Prod3 (MustNulls) (MayNulls) (Idx)
+
+  include HelperFunctionsIndexMustMaySets
+
+  let name () = "arrays containing null bytes"
+  type idx = Idx.t
+  type value = Val.t
+
+  type ret = Null | NotNull | Top
+
+  (* helper: returns Idx.maximal except for Overflows that are mapped to None *)
+  let idx_maximal i = match Idx.maximal i with
+    | Some i -> (try Some (Z.of_int (Z.to_int i)) with Z.Overflow -> None)
```

Comment:
What exactly is this doing? It goes from `Z -> int -> Z` ? If we need this, then

```suggestion
    | Some i when Z.fits_int i -> Some i
```

may be more appropriate. Also, if you want to make sure the value fits into `int`, why is the return type still `Z.t`? Something doesn't seem to quite add up here...

Reply:
`Idx.maximal` doesn't map $\infty$ to `None` but to an invalid value of `Z.t`, that whenever used causes the exception `Z.overflow`. Therefore I want to filter out such invalid values and set them to `None`. From my understanding, `Z.fits_int` also works. I think this wrapper makes the logic easier to understand in the rest of the code.